### ISSUE TITLE
datetime.datetime not supported

### ISSUE THREAD

**Issue by arthuston** [wolverdude/GenSON#26]
I am getting this error in genson:
genson.schema.node.SchemaGenerationError: Could not find matching type for object: datetime.datetime(2017, 1, 22, 15, 33, 50, 633000)

datetime.datetime is such a common datatype that I'm surprised it's not natively supported. Is there a work-around? I'm using GenSON 1.0.2 in PyCharm.

**Comment by wolverdude** [comment]
By the way, if you don't find this answer satisfactory, I've considered adding support for custom data types, which would allow you to support whatever data types you want, regardless of whether they fit the JSON spec or not.

I haven't added it because no one has actually requested it, but if there is a legitimate need, I'll be more likely to address it.

**Comment by arthuston** [comment]
Hi Jon,
Good answer, I'm aware of the JSON schema spec (and JSON itself) not
supporting datetime, and was also thinking more along the lines of
extensibility and custom data types. That would be a generally useful
feature, however, I have worked around this already and this is more of a
one-off project for me, so don't do it on my account.

Looks like a really neat and useful library that fits a real need - great
work and thank you.

**Comment by mkadaster** [comment]
Hi Jon,
Is this possible with the latest version in anyway? Can we add custom data types? Thanks.

**Comment by wolverdude** [comment]
Absolutely! See the docs [here](https://github.com/wolverdude/GenSON#customizing-schemabuilder). I would suggest building off of [`TypedSchemaStrategy`](https://github.com/wolverdude/GenSON#typedschemastrategy-api).

Here's an example of what your code might look like. I haven't tested it so no guarantees.

```python
import datetime
from genson import SchemaBuilder
from genson.schema.strategies import TypedSchemaStrategy

class DateTime(TypedSchemaStrategy):
    """ matches type: "datetime" (note: not JSON-Schema compliant) """
    JS_TYPE = 'datetime'
    PYTHON_TYPE = datetime.datetime

class DateTimeSchemaBuilder(SchemaBuilder):
    """ includes type: "datetime" (note: not JSON-Schema compliant) """
    EXTRA_STRATEGIES = (DateTime,)
```

**Comment by wolverdude** [comment]
The above is a non-compliant version that puts `"datetime"` in the `type` field. Alternately if you add some extra logic to your strategy to mark `datetime`s as strings with the extra keyword [`format: "date-time"`](https://json-schema.org/understanding-json-schema/reference/string.html#dates-and-times), you could create an implementation that also validates as a JSON-Schema.

**Comment by mkadaster** [comment]
Thanks for your feedback Jon. Yesterday I had walked through your documentation and noticed that this is possible indeed. So I came up with the below class extension and sharing it here incase anyone needs something like this as a sample in the future. Maybe we can also add it as part of your unit tests? Thanks.

`class CustomDateTime(TypedSchemaStrategy):
    """
    strategy for datetime values
    """
    JS_TYPE = 'string'
    PYTHON_TYPE = (str, type(u''))

    # create a new instance variable
    def __init__(self, node_class):
        super().__init__(node_class)
        self.format = "date-time"

    @classmethod
    def match_object(self, obj):
        super().match_object(obj)
        try:
            date_time_obj = datetime.datetime.strptime(obj, "%Y-%m-%dT%H:%M:%S.%f%z")
            if isinstance(date_time_obj, datetime.datetime):
                return True
            else:
                return False
        except (TypeError, ValueError) as exception:
            #print(exception)
            return False

    def to_schema(self):
        schema = super().to_schema()
        schema['type'] = self.JS_TYPE
        schema['format'] = self.format
        return schema`

`class CustomSchemaBuilder(SchemaBuilder):
    """ all number nodes include minimum """
    EXTRA_STRATEGIES = (CustomDateTime, )`

**Comment by mkadaster** [comment]
And the important point is ofcourse which I forgot to mention, this is json-schema compliant :)

**Comment by wolverdude** [comment]
Yup, this is exactly the type of situation that custom builders were implemented to support. Thanks for sharing!

**Comment by WilliamHYZhang** [comment]
Thanks! Very helpful
